gcs/src/main/scala/org/broadinstitute/monster/storage/gcs/GcsApi.scala

```diff
@@ -451,6 +449,89 @@ class GcsApi private[gcs] (runHttp: Request[IO] => Resource[IO, Response[IO]]) {
       }
     }
   }
+
+  /**
+    * List the content of a GCS bucket under a path prefix.
+    *
+    * Returns a failed Stream if no objects exist under the prefix.
+    *
+    * @param bucket name of the bucket to list contents within
+    * @param path prefix to match listed contents against
+    * @param pageSize max number of results to pull from GCS per HTTP request
+    */
+  def listContents(
+    bucket: String,
+    path: String,
+    pageSize: Int
+  ): Stream[IO, (String, FileType)] =
+    // State-tracking is non-obvious here.
```

Comment:
Note for anyone looking at this now: I just pushed a change which doesn't abuse `Option` for state-tracking. It should be simpler to read and review.